### ISSUE TITLE
finally found where the "generate a new label" assertion can fail / contains_eager() is doing nest selectable which it did not do before

### ISSUE THREAD

**Issue by sm-Fifteen** [sqlalchemy/sqlalchemy#8569]
### Describe the bug

When trying to generate the column names for an eager self-referential join, if the ORM model contains a `query_expression` or `column_property`, the query generation will fail with an `InvalidRequestError` as it fails to properly disambiguate the column key on the remote side. The second copy of each expression appears to use an anonymous table key instead of the aliased table name.

This seems to be related to #6090 and doesn't occur with SQLAlchemy 1.3.24 (it is a regression for my use case).

### To Reproduce

```python
from sqlalchemy.engine import create_engine
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.schema import Column
from sqlalchemy.types import TEXT, INTEGER
from sqlalchemy.orm import Session, relationship, remote, foreign, column_property, query_expression, aliased, contains_eager

from geoalchemy2 import Geometry
from geoalchemy2.functions import ST_AsGeoJSON
from geoalchemy2.functions import ST_Transform, ST_Intersects  # type:ignore[import]

MyORMBase = declarative_base(name='MyORMBase')


class Polygon(MyORMBase):
    __tablename__ = 'voronoi_poly'
    __table_args__ = {'schema': 'geodata'}

    objectid = Column(INTEGER, primary_key=True)
    entity_name = Column(TEXT, nullable=False)
    shape = Column(Geometry("POLYGON", srid=4326), nullable=False)

    # Conversion needs to be done on DB, the client doesn't understand the raw Geometry data.
    # These solutions seemingly don't work when ORM aliasing is involved
    # https://docs.sqlalchemy.org/en/14/orm/mapped_sql_expr.html
    shape_geojson_exp = query_expression(ST_AsGeoJSON(shape).label('shape_geojson_exp'))
    shape_geojson_cp = column_property(ST_AsGeoJSON(shape).label('shape_geojson_cp'))

    # https://docs.sqlalchemy.org/en/14/orm/join_conditions.html#custom-operators-based-on-sql-functions
    neighbors = relationship("Polygon", primaryjoin=ST_Intersects(shape, remote(foreign(shape))).as_comparison(1, 2), viewonly=True)


def fetch_voronoi_with_neighbors(sess_gisdata: Session) -> list[Polygon]:
    NeighborData = aliased(Polygon, name="voronoi_neighbors")

    voronoi_query = sess_gisdata.query(Polygon).join(
        # NeighborData is going to be the class referring to the contents of neighbors
        Polygon.neighbors.of_type(NeighborData)
    ).options(
        # Like a joinedload() (eager loading) but accounts for the alias and the .join
        # https://docs.sqlalchemy.org/en/13/orm/loading_relationships.html#contains-eager
        contains_eager(Polygon.neighbors, alias=NeighborData)
    ).distinct(NeighborData.entity_name)  # Don't report neighbors that have been listed for other initial polygons.

    voronoi_data = voronoi_query.all()

    return voronoi_data


def main():
    my_orm_engine = create_engine('sqlite://')  # Dialect or schema don't matter
    my_orm_session = Session(bind=my_orm_engine)

    try:
        # Fails with "no such table" on SQLA 1.3.24 (GeoA 0.8.5), that's ok
        # Fails with "Label name shape_geojson_exp is being renamed to an anonymous label due to disambiguation which is not supported right now." on SQLA 1.4.41 (GeoA 0.12.5)
        data = fetch_voronoi_with_neighbors(my_orm_session)
        print(data)
    finally:
        my_orm_session.close()
        my_orm_engine.dispose()


if __name__ == '__main__':
    main()
```

### Error

```
Traceback (most recent call last):
  File "sqla14_expr_attr.py", line 64, in <module>
    main()
  File "sqla14_expr_attr.py", line 56, in main
    data = fetch_voronoi_with_neighbors(my_orm_session)
  File "sqla14_expr_attr.py", line 44, in fetch_voronoi_with_neighbors
    voronoi_data = voronoi_query.all()
  File "lib\site-packages\sqlalchemy\orm\query.py", line 2772, in all
    return self._iter().all()
  File "lib\site-packages\sqlalchemy\orm\query.py", line 2907, in _iter
    result = self.session.execute(
  File "lib\site-packages\sqlalchemy\orm\session.py", line 1712, in execute
    result = conn._execute_20(statement, params or {}, execution_options)
  File "lib\site-packages\sqlalchemy\engine\base.py", line 1705, in _execute_20
    return meth(self, args_10style, kwargs_10style, execution_options)
  File "lib\site-packages\sqlalchemy\sql\elements.py", line 333, in _execute_on_connection
    return connection._execute_clauseelement(
  File "lib\site-packages\sqlalchemy\engine\base.py", line 1564, in _execute_clauseelement
    compiled_sql, extracted_params, cache_hit = elem._compile_w_cache(
  File "lib\site-packages\sqlalchemy\sql\elements.py", line 531, in _compile_w_cache
    compiled_sql = self._compiler(
  File "lib\site-packages\sqlalchemy\sql\elements.py", line 566, in _compiler
    return dialect.statement_compiler(dialect, self, **kw)
  File "lib\site-packages\sqlalchemy\sql\compiler.py", line 790, in __init__
    Compiled.__init__(self, dialect, statement, **kwargs)
  File "lib\site-packages\sqlalchemy\sql\compiler.py", line 463, in __init__
    self.string = self.process(self.statement, **compile_kwargs)
  File "lib\site-packages\sqlalchemy\sql\compiler.py", line 498, in process
    return obj._compiler_dispatch(self, **kwargs)
  File "lib\site-packages\sqlalchemy\sql\visitors.py", line 82, in _compiler_dispatch
    return meth(self, **kw)
  File "lib\site-packages\sqlalchemy\sql\compiler.py", line 3301, in visit_select
    compile_state = select_stmt._compile_state_factory(
  File "lib\site-packages\sqlalchemy\sql\base.py", line 510, in create_for_statement
    return klass.create_for_statement(statement, compiler, **kw)
  File "lib\site-packages\sqlalchemy\orm\context.py", line 765, in create_for_statement
    SelectState.__init__(self, self.statement, compiler, **kw)
  File "lib\site-packages\sqlalchemy\sql\selectable.py", line 4395, in __init__
    self.columns_plus_names = statement._generate_columns_plus_names(True)
  File "lib\site-packages\sqlalchemy\sql\selectable.py", line 6204, in _generate_columns_plus_names
    cols = self._all_selected_columns
  File "lib\site-packages\sqlalchemy\util\langhelpers.py", line 1184, in __get__
    obj.__dict__[self.__name__] = result = self.fget(obj)
  File "lib\site-packages\sqlalchemy\sql\selectable.py", line 6187, in _all_selected_columns
    return list(meth(self))
  File "lib\site-packages\sqlalchemy\sql\selectable.py", line 4620, in all_selected_columns
    return [c for c in _select_iterables(statement._raw_columns)]
  File "lib\site-packages\sqlalchemy\sql\base.py", line 93, in _select_iterables
    [c._select_iterable for c in elements]
  File "lib\site-packages\sqlalchemy\sql\base.py", line 93, in <listcomp>
    [c._select_iterable for c in elements]
  File "lib\site-packages\sqlalchemy\util\langhelpers.py", line 1113, in __get__
    obj.__dict__[self.__name__] = result = self.fget(obj)
  File "lib\site-packages\sqlalchemy\sql\selectable.py", line 737, in columns
    self._populate_column_collection()
  File "lib\site-packages\sqlalchemy\sql\selectable.py", line 1643, in _populate_column_collection
    self.element._generate_fromclause_column_proxies(self)
  File "lib\site-packages\sqlalchemy\sql\selectable.py", line 6362, in _generate_fromclause_column_proxies
    prox = [
  File "lib\site-packages\sqlalchemy\sql\selectable.py", line 6363, in <listcomp>
    c._make_proxy(
  File "lib\site-packages\sqlalchemy\sql\elements.py", line 4640, in _make_proxy
    raise exc.InvalidRequestError(
sqlalchemy.exc.InvalidRequestError: Label name shape_geojson_exp is being renamed to an anonymous label due to disambiguation which is not supported right now.  Please use unique names for explicit labels.

Process finished with exit code 1
``` 

### Versions

- OS: Windows 10
- Python: 3.10.7
- SQLAlchemy: 1.4.41
- Database: PostgreSQL 14 (Minimal Code Example uses SQLite)
- DBAPI: psycopg (Example uses pysqlite)

### Additional context

I know the MCVE is a bit long and convoluted, but it's a niche use case that's hard to make sense of when reduced into primitive values. The obvious `hybrid_property` solution isn't an option here because the Python side of things wouldn't understand the raw Geometry data.

**Comment by zzzeek** [comment]
for the test case, just replace the Geo functions with a regular SQL function.   

there are two separate issues here, one is non-controversial which is the labeling error, the other is that it's trying to do an joined eagerloading subquery here which was not previously part of the contract of contains_eager(), this was unexpected but if I change it now, it will change people's queries quite dramatically.

so first you want to confirm your problems are completely solved with this part:

1. fix your models to not have an explicit label name, which doesn't make much sense to have:

```py
    shape_geojson_exp = query_expression(ST_AsGeoJSON(shape))
    shape_geojson_cp = column_property(ST_AsGeoJSON(shape))
```


2. with that change, we just skip the error here:

```diff
diff --git a/lib/sqlalchemy/sql/elements.py b/lib/sqlalchemy/sql/elements.py
index 268c0d6ac4..ace43b3a1d 100644
--- a/lib/sqlalchemy/sql/elements.py
+++ b/lib/sqlalchemy/sql/elements.py
@@ -4636,7 +4636,7 @@ class Label(roles.LabeledColumnExprRole, ColumnElement):
         # when a label name conflicts with other columns and select()
         # is attempting to disambiguate an explicit label, which is not what
         # the user would want.   See issue #6090.
-        if key != self.name:
+        if key != self.name and not isinstance(self.name, _anonymous_label):
             raise exc.InvalidRequestError(
                 "Label name %s is being renamed to an anonymous label due "
                 "to disambiguation "

```

however the distinct() is still going to give you a more complex query:

```sql
SELECT anon_1."ST_AsGeoJSON_1" AS "anon_1_ST_AsGeoJSON_1", anon_1."ST_AsGeoJSON_2" AS "anon_1_ST_AsGeoJSON_2", anon_1."ST_AsGeoJSON_1_1" AS "anon_1_ST_AsGeoJSON_1_1", anon_1."ST_AsGeoJSON_2_2" AS "anon_1_ST_AsGeoJSON_2_2", anon_1.voronoi_neighbors_objectid AS anon_1_voronoi_neighbors_objectid, anon_1.voronoi_neighbors_entity_name AS anon_1_voronoi_neighbors_entity_name, AsEWKB(anon_1.voronoi_neighbors_shape) AS anon_1_voronoi_neighbors_shape, anon_1.geodata_voronoi_poly_objectid AS anon_1_geodata_voronoi_poly_objectid, anon_1.geodata_voronoi_poly_entity_name AS anon_1_geodata_voronoi_poly_entity_name, AsEWKB(anon_1.geodata_voronoi_poly_shape) AS anon_1_geodata_voronoi_poly_shape 
FROM (SELECT DISTINCT AsGeoJSON(geodata.voronoi_poly.shape) AS "ST_AsGeoJSON_1", AsGeoJSON(geodata.voronoi_poly.shape) AS "ST_AsGeoJSON_2", AsGeoJSON(voronoi_neighbors.shape) AS "ST_AsGeoJSON_1", AsGeoJSON(voronoi_neighbors.shape) AS "ST_AsGeoJSON_2", voronoi_neighbors.objectid AS voronoi_neighbors_objectid, voronoi_neighbors.entity_name AS voronoi_neighbors_entity_name, voronoi_neighbors.shape AS voronoi_neighbors_shape, geodata.voronoi_poly.objectid AS geodata_voronoi_poly_objectid, geodata.voronoi_poly.entity_name AS geodata_voronoi_poly_entity_name, geodata.voronoi_poly.shape AS geodata_voronoi_poly_shape 
FROM geodata.voronoi_poly JOIN geodata.voronoi_poly AS voronoi_neighbors ON ST_Intersects(geodata.voronoi_poly.shape, voronoi_neighbors.shape)) AS anon_1
```

the next part which I have to see how this happened, why it wasn't noticed, and if it's going to be a big deal, is this:

```diff
diff --git a/lib/sqlalchemy/orm/strategies.py b/lib/sqlalchemy/orm/strategies.py
index 944c114a64..6df34d5496 100644
--- a/lib/sqlalchemy/orm/strategies.py
+++ b/lib/sqlalchemy/orm/strategies.py
@@ -1949,8 +1949,6 @@ class JoinedLoader(AbstractRelationshipLoader):
 
         if not compile_state.compile_options._enable_eagerloads:
             return
-        elif self.uselist:
-            compile_state.multi_row_eager_loaders = True
 
         path = path[self.parent_property]
 
@@ -2002,6 +2000,9 @@ class JoinedLoader(AbstractRelationshipLoader):
                 chained_from_outerjoin,
             )
 
+            if self.uselist:
+                compile_state.multi_row_eager_loaders = True
+
         with_poly_entity = path.get(
             compile_state.attributes, "path_with_polymorphic", None
         )

```

then you get the query that should be the same as 1.3

```sql
 SELECT DISTINCT AsGeoJSON(geodata.voronoi_poly.shape) AS "ST_AsGeoJSON_1", AsGeoJSON(geodata.voronoi_poly.shape) AS "ST_AsGeoJSON_2", AsGeoJSON(voronoi_neighbors.shape) AS "ST_AsGeoJSON_1", AsGeoJSON(voronoi_neighbors.shape) AS "ST_AsGeoJSON_2", voronoi_neighbors.objectid AS voronoi_neighbors_objectid, voronoi_neighbors.entity_name AS voronoi_neighbors_entity_name, AsEWKB(voronoi_neighbors.shape) AS voronoi_neighbors_shape, geodata.voronoi_poly.objectid AS geodata_voronoi_poly_objectid, geodata.voronoi_poly.entity_name AS geodata_voronoi_poly_entity_name, AsEWKB(geodata.voronoi_poly.shape) AS geodata_voronoi_poly_shape 
FROM geodata.voronoi_poly JOIN geodata.voronoi_poly AS voronoi_neighbors ON ST_Intersects(geodata.voronoi_poly.shape, voronoi_neighbors.shape)
```


I need confirmation on patch A and B separately, thanks

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **main** branch:

**accommodate labeled exprs in contains_eager / repair nesting** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4094

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **rel_1_4** branch:

**accommodate labeled exprs in contains_eager / repair nesting** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4095

**Comment by zzzeek** [comment]
test case

```py
from sqlalchemy import Column
from sqlalchemy import ForeignKey
from sqlalchemy import func
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import aliased
from sqlalchemy.orm import column_property
from sqlalchemy.orm import contains_eager
from sqlalchemy.orm import relationship
from sqlalchemy.orm import Session

Base = declarative_base()


class A(Base):
    __tablename__ = "a"

    id = Column(Integer, primary_key=True)
    a_id = Column(ForeignKey("a.id"))
    data = Column(String)
    data_lower = column_property(func.lower(data))

    bs = relationship("A")


s = Session()

ba = aliased(A)
print(
    s.query(A)
    .join(A.bs.of_type(ba))
    .options(contains_eager(A.bs.of_type(ba)))
    .distinct(ba.data)
)

```

**Comment by zzzeek** [comment]
Hm here's a completely different patch for that first part, I need to look at some history to see what should be used

```diff
diff --git a/lib/sqlalchemy/orm/strategies.py b/lib/sqlalchemy/orm/strategies.py
index 944c114a64..e19361d254 100644
--- a/lib/sqlalchemy/orm/strategies.py
+++ b/lib/sqlalchemy/orm/strategies.py
@@ -2111,7 +2111,7 @@ class JoinedLoader(AbstractRelationshipLoader):
                 user_defined_adapter,
             )
 
-        add_to_collection = context.primary_columns
+        add_to_collection = context.secondary_columns
         return user_defined_adapter, adapter, add_to_collection
 
     def _gen_pooled_aliased_class(self, context):

```

**Comment by zzzeek** [comment]
OK whether that part goes to "secondary_columns" or not is based on if we apply compound wrapping for contains_eager() or not.  contains_eager() wasn't intended for compound wrapping so we likely need to just fix that part

**Comment by sm-Fifteen** [comment]
> 1. fix your models to not have an explicit label name, which doesn't make much sense to have:
> 
> 
> ```python
>     shape_geojson_exp = query_expression(ST_AsGeoJSON(shape))
>     shape_geojson_cp = column_property(ST_AsGeoJSON(shape))
> ```

I didn't have those initially, but they made it easier to keep track of things while stepping through SQLA code.

> I need confirmation on patch A and B separately, thanks

From me, or is this a keyphrase for Gerrit?

**Comment by zzzeek** [comment]
I was looking for testing help but since then things have moved fast.     You can leave the labels in there, pretty much 1.3's approach is restored

**Comment by zzzeek** [comment]
when this issue is closed (and 1.4.42 is released or you work from SQLA rel_1_4 branch in github)
